### DISCUSSION TITLE
ImageMagick: Changing background in tif-files and convert to JPEG

### DISCUSSION THREAD

**Discussion by JesperArentoft** [ImageMagick/ImageMagick#6360] [Help] [answered]
### ImageMagick version

7.1.0-62

### Operating system, version and so on

Windows

### Description

Hi all. 

I have a number of tif files with transparent background that need to be converted to JPEG with a gray background. I have tried different combinations, but unfortunately nothing works. I just get a white background.

I have among other things tried with the following string: 
convert input_image.tif -flatten -background gray output_image.jpg

I have tried to look for solutions, but nothing points in this direction. 

Do you know if it is possible to change the background color in tif files with ImageMagick?

Best regards
Jesper A.

**Comment by JesperArentoft** [comment]
Here is a test file where I can't change the background when converting to JPEG.
[test1.tif.zip](https://github.com/ImageMagick/ImageMagick/files/11564473/test1.tif.zip)

**Comment by snibgo** [comment]
You have IM v7. The usual command for that is `magick`. But you use `convert` which uses the very old v6 syntax, and is provided for backwards compatibility. The v7 syntax is better. I suggest using `magick`. (And not "`magick convert`" which would also use v6 syntax.)

IM processes options in the order you give them.

`-flatten` and "`-layers flatten`" will use the current background colour. You set this, but only *after* you flatten. That is too late. You need to set it *before* doing the flatten.

**Reply by JesperArentoft** [reply]
Dear Snibgo

Thank you for the advice. I appreciate the tip and appreciate our help :-)

Best regards
Jesper

**Comment by JesperArentoft** [comment]
Hi again

It works for some files, but not with the uploaded test file (test1.tif.zip). 

The simple ImageMagick string that I have tested with is: magick input.tif -background blue -flatten output.jpg, but this just give a white background and not the blue one.

Could there be any special characteristics in this particular file that require additional commands?

**Comment by snibgo** [comment, marked as answer]
```
D:\temp>%IMG7%magick c:\web\im\test1.tif info:

magick: Unknown field with tag 40965 (0xa005) encountered. `TIFFReadCustomDirectory' @ warning/tiff.c/TIFFWarnings/951.
magick: Unknown field with tag 59932 (0xea1c) encountered. `TIFFReadCustomDirectory' @ warning/tiff.c/TIFFWarnings/951.
magick: Unknown field with tag 59933 (0xea1d) encountered. `TIFFReadCustomDirectory' @ warning/tiff.c/TIFFWarnings/951.
c:\web\im\test1.tif[0] TIFF 2000x3000 2000x3000+0+0 8-bit TrueColor sRGB 19.9199MiB 0.000u 0:00.000
c:\web\im\test1.tif[1] TIFF 1560x234 1560x234+230+2429 8-bit sRGB 19.9199MiB 0.000u 0:00.003
c:\web\im\test1.tif[2] TIFF 1695x2282 1695x2282+152+337 8-bit sRGB 19.9199MiB 0.000u 0:00.004
```
Don't worry about the warnings.

Your tiff file contains three images, so your command creates an output tiff with three images. Is that what you want?

Adding `-verbose` tells me that only the third image has an alpha channel. Hence, flattening the images against a blue background affects only the third image. It gets a blue background, as expected.

All three images have a clipping path. By default, IM ignores these. But you can use clipping paths, like this:
```
magick c:\web\im\test1.tif[0] -alpha transparent -clip -alpha opaque +clip -background Blue -layers Flatten x.tiff
```
This reads just the first image, and makes it fully transparent. `-clip` says to use the clipping path for future operations. `-alpha opaque` makes pixels inside the clipping path transparent. `+clip` stops using the clipping path. Then we flatten against blue, and save the image to a file.

Does that help?

**Reply by JesperArentoft** [reply]
Hi Snibgo

It definitely helps. 

I have tested with a number of files and here the result is what we want. Now I will check if it works on all of them. 

You have been a huge help - 10000 thanks :-)
